### PR TITLE
Changes the labels used by Juju for Kubernetes.

### PR DESCRIPTION
**Pull request by tlm**
Juju now adopts more idiomatic way of adding Kubernetes labels as per
k8s docs
https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/

We are adopting k8s standard labels and also adding juju.is labels such
as model.juju.is/name

This change will not retroactively change labels for existing models

### Checklist

 - [ ] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [ ] Do comments answer the question of why design decisions were made?

----

## Description of change

*Please replace with a description about why this change is needed, along with a description of what changed?*

## QA steps

*Please replace with how we can verify that the change works?*

```sh
QA steps here
```

## Documentation changes

*Please replace with any notes about how it affects current user workflow? CLI? API?*

## Bug reference

*Please add a link to any bugs that this change is related to.*

### CODE REVIEW

File: caas/kubernetes/provider/annotations.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	// legacyAnnotationStorageName is the legacy annotation used by Juju for
+	// dictating storage name on k8s storage objects.
+	legacyAnnotationStorageName = "juju-storage"
+
+	// legacyAnnotationVersion is the legacy annotation used by Juju for
+	// dictating juju agent version on operators.
+	legacyAnnotationVersion = "juju-version"
+)
+
+type AnnotationKeySupplier func() string
```
Comment (wallyworld) [inline caas/kubernetes/provider/annotations.go:28]: this doesn't seem to be used anywhere?

File: caas/kubernetes/provider/annotations.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+package provider
+
+import (
+	"github.com/juju/juju/core/annotations"
+)
+
+const (
+	// AnnotationJujuStorageName is the Juju annotation that represents a
+	// storage objects associated Juju name.
+	AnnotationJujuStorageName = "storage.juju.is/name"
```
Comment (wallyworld) [inline caas/kubernetes/provider/annotations.go:13]: I don't think these need to be exported do they?

File: caas/kubernetes/provider/annotations.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	legacyAnnotationVersion = "juju-version"
+)
+
+type AnnotationKeySupplier func() string
+
+// AnnotationsForStorage provides the annotations that should be placed on a
+// storage object. The annotations returned by this function are storage
+// specific only and should be combined with other annotations where
+// appropriate.
+func AnnotationsForStorage(name string, legacy bool) annotations.Annotation {
```
Comment (wallyworld) [inline caas/kubernetes/provider/annotations.go:34]: Same with these - they are only used in the provider package

File: caas/kubernetes/provider/labels.go [hunk 4]
```diff
@@ -3,44 +3,201 @@
...
+	for k, v := range has {
+		if src[k] != v {
+			return false
+		}
+	}
+	return true
+}
+
+// LabelSetToSelector converts a set of Kubernetes labels
+func LabelSetToSelector(l labels.Set) labels.Selector {
```
Comment (wallyworld) [inline caas/kubernetes/provider/labels.go:101]: These don't need to be exported either
Well, LabelsForModel does, but a bunch of others no

File: caas/kubernetes/provider/labels.go [hunk 5]
```diff
@@ -3,44 +3,201 @@
...
+	}
+}
+
+func LabelForKeyValue(key, value string) labels.Set {
+	return labels.Set{
+		key: value,
+	}
+}
+
+// LabelsMerge
```
Comment (wallyworld) [inline caas/kubernetes/provider/labels.go:165]: need to complete the comment, but I don't think this need to be exported anyway

File: caas/kubernetes/provider/labels.go [hunk 6]
```diff
@@ -3,44 +3,201 @@
...
+		return false, nil
+	}
+	if err != nil {
+		return true, errors.Annotatef(err, "unable to determine legacy status for namespace %s", model)
+	}
+
+	return !HasLabels(ns.Labels, LabelsForModel(model, false)), nil
+}
+
+func StorageNameFromLabels(labels labels.Set) string {
```
Comment (wallyworld) [inline caas/kubernetes/provider/labels.go:198]: comment exported functions, unless they do not really need to be exported

File: caas/kubernetes/provider/labels_test.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func (l *LabelSuite) TestIsLegacyModelLabels(c *gc.C) {
+	tests := []struct {
+		IsLegacy  bool
+		Model     string
+		Namespace *core.Namespace
+	}{
+		{
+			IsLegacy: false,
```
Comment (wallyworld) [inline caas/kubernetes/provider/labels_test.go:30]: there's only 1 test here and we have a slice of tests
i think you meant to add another test with IsLegacy: true

**Review by wallyworld — APPROVED** [review]
A meta comment is that there's a bunch of stuff exported that doesn't need to be.
There also appears to be a lot of places where code is changed to add the new labels but little in the way of corresponding unit test changes. Are we missing unit tests to check for labels?

File: caas/kubernetes/provider/annotations.go (hunk 3 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/annotations.go:34]: we could make `AnnotationsForStorage` and `AnnotationsForVersion` consistent

File: caas/kubernetes/provider/labels.go [hunk 8]
```diff
@@ -3,44 +3,201 @@
...
-			dest[k] = v
-		}
+import (
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	core "k8s.io/client-go/kubernetes/typed/core/v1"
+
```
Comment (ycliuhw) [inline caas/kubernetes/provider/labels.go:13]: no space here

File: caas/kubernetes/provider/operator.go [hunk 9]
```diff
@@ -132,13 +139,27 @@ func (k *kubernetesClient) ensureOperatorRBACResources(operatorName string, labe
 
 // EnsureOperator creates or updates an operator pod with the given application
 // name, agent path, and operator config.
-func (k *kubernetesClient) EnsureOperator(appName, agentPath string, config *caas.OperatorConfig) (err error) {
-	logger.Debugf("creating/updating %s operator", appName)
+func (k *kubernetesClient) EnsureOperator(name, agentPath string, config *caas.OperatorConfig) (err error) {
+	logger.Debugf("creating/updating %s operator", name)
```
Comment (ycliuhw) [inline caas/kubernetes/provider/operator.go:143]: appName is more accurate, right?
ppl will think name here is the operator name

File: caas/kubernetes/provider/operator.go [hunk 10]
```diff
@@ -660,10 +682,8 @@ func operatorPod(
 	mountToken := true
 	return &core.Pod{
 		ObjectMeta: v1.ObjectMeta{
-			Name: podName,
-			Annotations: podAnnotations(annotations.Copy()).
-				Add(labelVersion, version).ToMap(),
-			Labels: operatorLabels(appName),
+			Annotations: podAnnotations(annotations.Copy()).ToMap(),
```
Comment (ycliuhw) [inline caas/kubernetes/provider/operator.go:685]: we want to remove version from annotation?

File: caas/kubernetes/provider/rbac.go [hunk 11]
```diff
@@ -25,17 +25,17 @@ import (
 // given ServiceAccount. If app name cannot be obtained from the service
 // account, errors.NotFound is returned.
 func AppNameForServiceAccount(sa *core.ServiceAccount) (string, error) {
-	if appName, ok := sa.Labels[labelApplication]; ok {
+	if appName, ok := sa.Labels[LabelKubernetesAppName]; ok {
```
Comment (ycliuhw) [inline caas/kubernetes/provider/rbac.go:28]: do we want to support legacy models?

File: caas/kubernetes/provider/rbac.go [hunk 12]
```diff
@@ -330,16 +334,18 @@ func (k *kubernetesClient) deleteRole(name string, uid types.UID) error {
...
-		return nil
+func (k *kubernetesClient) deleteRoles(selectors ...k8slabels.Selector) error {
+	for _, selector := range selectors {
+		err := k.client().RbacV1().Roles(k.namespace).DeleteCollection(&v1.DeleteOptions{
+			PropagationPolicy: &defaultPropagationPolicy,
+		}, v1.ListOptions{
+			LabelSelector: selector.String(),
+		})
+		if !k8serrors.IsNotFound(err) {
+			return errors.Trace(err)
```
Comment (ycliuhw) [inline caas/kubernetes/provider/rbac.go:345]: do we want to continue if err == nil?

File: caas/kubernetes/provider/rbac.go [hunk 13]
```diff
@@ -542,16 +548,18 @@ func (k *kubernetesClient) deleteRoleBinding(name string, uid types.UID) error {
...
+func (k *kubernetesClient) deleteRoleBindings(selectors ...k8slabels.Selector) error {
+	for _, selector := range selectors {
+		err := k.client().RbacV1().RoleBindings(k.namespace).DeleteCollection(&v1.DeleteOptions{
+			PropagationPolicy: &defaultPropagationPolicy,
+		}, v1.ListOptions{
+			LabelSelector: selector.String(),
+		})
+		if !k8serrors.IsNotFound(err) {
+			return errors.Trace(err)
+		}
```
Comment (ycliuhw) [inline caas/kubernetes/provider/rbac.go:560]: do we want to continue if err == nil?

File: worker/caasadmission/handler.go [hunk 14]
```diff
@@ -171,15 +171,17 @@ func errToAdmissionResponse(err error) *admission.AdmissionResponse {
...
 
-func patchForLabels(labels map[string]string, appName string) []patchOperation {
+func patchForLabels(
+	labels map[string]string,
+	appName string,
+	legacyLabels bool) []patchOperation {
 	patches := []patchOperation{}
 
-	neededLabels := provider.LabelsForApp(appName)
+	neededLabels := provider.LabelForKeyValue(provider.LabelJujuAppCreatedBy, appName)
```
Comment (ycliuhw) [inline worker/caasadmission/handler.go:184]: do we need `legacyLabels` here?

**Review by ycliuhw — COMMENTED** [review]
just a few small comments.

**Comment by wallyworld** [comment]
[standardise juju labels and annotations](https://trello.com/c/xdY2LsO9/1131-standardise-juju-labels-and-annotations)

**Pull request closed without merging**